docs: agent refactor (#8524)
langflow-ai/langflow: docs/docs/Agents/agents.md

---
title: Langflow Agents
slug: /agents
---

import Icon from "@site/src/components/icon";

**Agents** use LLMs as a brain to autonomously analyze problems and select tools to solve them.

To simplify the development of agents, Langflow created an [Agent](/components-agents#agent-component) component that simplifies configuration and lets developers focus on solving problems.

The Agent component provides everything needed to create an agent, including multiple LLM models, a built-in instructions field, and tool connection and management.

Learn how to build a flow starting with the **Agent** component, and see how it can help you solve problems.

## Prerequisites

> - [An OpenAI API key](https://platform.openai.com/)

Reviewer: This is only if you choose OpenAI as the provider. I think I would break the basic config and demo apart from the tool config:

agents.md
```
# Use Langflow Agents

<introduce the Agent component>

## Agent settings
<keep this content as minimal as possible>

You can configure the Agent component to use your preferred provider and model, custom instructions, and tools.

### Agent models and providers

Use the *Model Provider* and *Model Name* settings to select the LLM that you want the Agent to use.

You must provide an authentication key for the selected provider, such as an OpenAI API key for OpenAI models.

### Agent instructions and input

In the *Agent Instructions* field, you can provide custom instructions that you want the Agent to use for every conversation.

These instructions are applied in addition to the *Input*, which is provided at runtime.

### Agent tools

Agents are most useful when they have the appropriate tools available to complete requests.

An Agent component can use any Langflow component as a tool, as long as you attach it to the Agent component.

When you attach a component as a tool, you must configure the component as a tool.
For more information and instructions, see [Configure tools for agents](agent-tools.md).

## Use the Agent component in a flow

<I think this title is more clear, because "create a flow with the agent compnent" could be misinterpreted as the component building the flow for you>

The following steps demonstrate how to use the Agent component by building a problem-solving agent flow.

<instructions for creating the flow>
<I think you can merge the prerequisite with step 3, like: Use the default model or select another provider and model. If you want to use the default model, you need an OpenAI API key to run the flow.>
<In LF Desktop, I noticed this Agent flow is already build in as a starter flow. Maybe these instructions could be more generic and you could have a TIP indicating that they can use the Agent starter flow to see a prebuild demo?>

## See also

- [agent-tools.md]
```

agent-tools.md
```
<move all details about tool configuration to this page>
```

Author: Great idea, thanks. Added

## Use an agent as a tool

The agent component itself also supports **Tool Mode** for creating multi-agent flows.

Add an agent to your flow that uses a different OpenAI model for a larger context window.

1. Click and drag an **Agent** component to your workspace.
2. Add your **Open AI API Key** to the **Agent** component.
3. In the **Model Name** field, select `gpt-4.1`.
4. Click **Tool Mode** to use this new agent as a tool.
5. Connect the new agent's **Toolset** port to the previously created agent's **Tools** port.
6. Connect the **Web Search**, **URL**, and **Calculator** to the new agent.
The new agent will use `gpt-4.1` for the larger tasks of scraping and searching information that require large context windows.
The previously created agent will now use this agent as a tool, with its unique LLM and toolset.

![Agent as a tool](/img/agent-example-agent-as-tool.png)

7. The new agent's actions can be edited to help the agent understand how to use it.
Click <Icon name="SlidersHorizontal" aria-hidden="true"/> **Edit Tools** to modify its `name`, `description`, or `enabled` metadata.

> For example, the default tool name is `Agent`. Edit the name to `Agent-gpt-41`, and edit the description to `Use the gpt-4.1 model for complex problem solving`. The connected agent will understand that this is the `gpt-4.1` agent, and will use it for tasks requiring a larger context window.

Author: @edwinjosechittilappilly This is the bug that I reported, where the connected agent as tool displays potentially incorrect information.

            "reversed_text": reversed_text,
            "uppercase_text": uppercase_text
        }

        data = Data(value=analysis_result)
        self.status = data
        return data
```
</details>

3. To use the custom component as a tool, click **Tool Mode**.
4. Connect the custom component's tool output to the agent's tools input.
5. Open the Playground and instruct the agent, `Use the text analyzer on this text: "Agents really are thinking machines!"`

<details open>
<summary>Response</summary>
```
AI
gpt-4o
Finished
0.6s
Here is the analysis of the text "Agents really are thinking machines!":
Original Text: Agents really are thinking machines!
Word Count: 5
Character Count: 36
Sentence Count: 1
Reversed Text: !senihcam gnikniht era yllaer stnegA
Uppercase Text: AGENTS REALLY ARE THINKING MACHINES!
```
</details>

The agent correctly calls the `analyze_text` action and returns the result to the Playground.

## Make any component a tool

If the component you want to use as a tool doesn't have a **Tool Mode** button, add `tool_mode=True` to one of the component's inputs, and connect the new **Toolset** output to the agent's **Tools** input.

> Langflow supports **Tool Mode** for the following data types:

Author: @edwinjosechittilappilly Confirm this please

---
title: Use Langflow Agents
slug: /agents
---

import Icon from "@site/src/components/icon";

Agents use LLMs as a brain to autonomously analyze problems and select tools to solve them.

Langflow's [Agent component](/components-agents#agent-component) simplifies agent configuration so you can focus on application development.

The Agent component provides everything you need to create an agent, including multiple LLMs, custom instructions, and tool configuration.

## Agent settings

You can configure the Agent component to use your preferred provider and model, custom instructions, and tools.

### Agent models and providers

Use the *Model Provider* and *Model Name* settings to select the LLM that you want the Agent to use.

You must provide an authentication key for the selected provider, such as an OpenAI API key for OpenAI models.

### Agent instructions and input

> In the *Agent Instructions* field, you can provide custom instructions that you want the Agent to use for every conversation.

Reviewer: ```suggestion
In the **Agent Instructions** field, you can provide custom instructions that you want the Agent component to use for every conversation.
```

> These instructions are applied in addition to the *Input*, which is provided at runtime.

Reviewer: ```suggestion
These instructions are applied in addition to the **Input**, which is provided at runtime.
```

### Agent tools

Agents are most useful when they have the appropriate tools available to complete requests.

> An Agent component can use any Langflow component as a tool, as long as you attach it to the Agent component.

Reviewer: ```suggestion
An Agent component can use any Langflow component as a tool, as long as you attach it to the Agent component.

:::tip
To allow agents to use tools from MCP servers, use the [**MCP connection** component](/components-tools#mcp-connection). 
:::
```

> When you attach a component as a tool, you must configure the component as a tool by enabling *Tool Mode*.

Reviewer: ```suggestion
When you attach a component as a tool, you must configure the component as a tool by enabling **Tool Mode**.
```

For more information, see [Configure tools for agents](/)

## Use the Agent component in a flow

:::tip

> For a pre-built demonstration, open the **Simple Agent** starter flow and follow along.

Reviewer: ```suggestion
For a pre-built demonstration, open the **Simple Agent** template flow and follow along.
```

Are we supposed to just call them templates?

:::

Create an agent in Langflow, starting with the **Agent** component and working outward.

1. Click **New Flow**, and then click **Blank Flow**.

> 2. Click and drag an **Agent** component to your workspace.

Reviewer: ```suggestion
2. Add an **Agent** component to your workspace.
```

I think add is fine cause you can drag or hit the +.

> 3. Use the default model or select another provider and model. If you want to use the default model, you need an OpenAI API key to run the flow.

Reviewer: ```suggestion
3. Use the default model or select another provider and model, and then provide credentials for your chosen provider. For example, to use the default model, you must provide an OpenAI API key.
```

4. Add **Chat input** and **Chat output** components to your flow, and connect them to the tool calling agent.

> ![Chat with agent component](/img/agent-example-add-chat.png)
> 
> This basic flow allows you to chat with the agent in the **Playground**, but you're only chatting with the OpenAI LLM.
> To unlock the power of the Agent component, connect some tools.

Reviewer: ```suggestion
    ![Chat with agent component](/img/agent-example-add-chat.png)

    This basic flow allows you to chat with the agent in the **Playground**, but you're only chatting with the selected LLM.
    To unlock the power of the Agent component, connect some tools.
```

Should the playground icon be here?

5. Add the **News Search**, **URL**, and **Calculator** components to your flow.
6. Enable **Tool Mode** in the **News Search**, **URL**, and **Calculator** components.

> To enable **Tool Mode** in a component, in the component's configuration tab, click **Tool Mode**.

Reviewer: 59 and 60 are a bit redundant. In the MCP PR, I kind of avoided this awkward instruction by linking to the Component overview page where it introduces the component menu/tab.

Author: Good idea, added

> **Tool Mode** makes a component into a tool by adding a **Toolset** port that can be connected to an agent's **Tools** port.
> Enabling **Tool Mode** modifies the component's inputs to accept requests from the Agent to use a tool's available actions.

Reviewer: ```suggestion

    **Tool Mode** makes a component into a tool by modifying the component's inputs to accept requests from the Agent component to use a tool's available actions. A component in tool mode has a **Toolset** port that you must connect to an Agent component's **Tools** port if you want to allow the agent to use the tool's actions.
```

The step is a bit text heavy as one single block.

Also a technicality, but the modification of the inputs is, IMO what "makes" it into a tool. The appearance of the toolset port is a side effect.

> 7. Connect the **News Search**, **URL**, and **Calculator** tool components to your agent.

Reviewer: ```suggestion
7. Connect the **Toolset** port on the three tool components to the **Tools** port on the Agent component.
```

optional but to clarify the ports being connected

![Chat with agent component](/img/agent-example-add-tools.png)

> 8. Open the **Playground**. Ask the agent, `What tools are you using to answer my questions?`

Reviewer: Should the icon be here?

The agent should respond with a list of the connected tools.

```text
I use a combination of my built-in knowledge (up to June 2024) and a set of external tools to answer your questions. Here are the main types of tools I can use:
Web Search & Content Fetching: I can fetch and summarize content from web pages, including crawling links recursively.
News Search: I can search for recent news articles using Google News via RSS feeds.
Calculator: I can perform arithmetic calculations and evaluate mathematical expressions.
Date & Time: I can provide the current date and time in various time zones.
These tools help me provide up-to-date information, perform calculations, and retrieve specific data from the internet when needed. If you have a specific question, let me know, and I’ll use the most appropriate tool(s) to help!
```

9. Ask the agent, `Summarize today's tech news`.

> The Playground displays the agent's tool calls, what input was provided, and the raw output the agent received before performing the summary. The agent should call the News Search component's `search_news` action.

Reviewer: ```suggestion
The Playground displays the agent's tool calls, what input was provided, and the raw output the agent received before generating the summary. The agent should call the **News Search** component's `search_news` action.
```

---
title: Use Langflow Agents
slug: /agents
---

import Icon from "@site/src/components/icon";

Agents use LLMs as a brain to autonomously analyze problems and select tools to solve them.

Langflow's [Agent component](/components-agents#agent-component) simplifies agent configuration so you can focus on application development.

> The Agent component provides everything you need to create an agent, including multiple LLMs, custom instructions, and tool configuration.

Reviewer 2: Tool configurations?

Author: Cruft, fixed

> The Agent component provides everything you need to create an agent, including multiple LLMs, custom instructions, and tool configuration.

Author: ```suggestion
The Agent component provides everything you need to create an agent, including multiple LLM providers and custom instructions.
```
